src/ECS/Systems/PathfindingSystem.cpp

```diff
@@ -28,43 +39,557 @@ PathfindingSystem& PathfindingSystem::instance()
+					const auto t1 = angle1 * 2.0f / 3.0f * r1 / wallHug.speed;
+					int t = static_cast<int>(glm::round(glm::min(t0, t1)));
+
+					assert(t >= 0);
+					if (t < 1)
+					{
+						// We're too close to second circle. Act like we're on the second circle and continue looking forward by
+						// recursively calling function with new obstacle.
+						reference.entity = *iter;
+						OrbitScanForObstacle(entity, clockwise, transform, wallHug);
+					}
+					else if (t < 4)
+					{
+						t = 0;
+					}
+					else
+					{
+						t = glm::min(t, 255);
+					}
+
+					reference.stepsAway = static_cast<uint8_t>(t);
+				}
+				else
+				{
+					// TODO:
+					// if intersect[0].obj is None:  # True
+					//     self.init_steps_xz()
+					//     # self.field_0x78 = 0x10
+					//     self.circle_hug_info.reset(self)
+					//     self.move_state = MoveState.STEP_THROUGH
+					// assert(false);
+				}
+			}
+		}
+	}
+
+	InitializeStepAroundObstacle(transform, wallHug, obstacleFixed, numCirclesAway, clockwise);
+
+	return true;
+}
+
+template <MoveState S, typename... Exclude>
+void StepForward(ecs::Registry& registry, Exclude... exclude)
+{
+	registry.Each<MoveStateTagComponent<S>, const WallHug, Transform>(
+	    [](MoveStateTagComponent<S>& state, const WallHug& wallHug, const Transform& transform) {
+		    const auto goal = glm::xz(transform.position) + wallHug.step;
+		    state.stepGoal = goal;
+	    },
+	    exclude...);
+}
+
+template <MoveState S>
+bool CellTransition(entt::entity entity, const MoveStateTagComponent<S> state, Transform& transform, WallHug& wallHug);
+
+template <>
+bool CellTransition(entt::entity entity, [[maybe_unused]] const MoveStateTagComponent<MoveState::Linear> state,
+                    Transform& transform, WallHug& wallHug)
+{
+	InitializeStepToGoal(transform, wallHug);
+	return LinearScanForObstacle(entity, glm::xz(transform.position), wallHug.step);
+}
+
+template <>
+bool CellTransition(entt::entity entity, const MoveStateTagComponent<MoveState::Orbit> state, Transform& transform,
+                    WallHug& wallHug)
+{
+	return OrbitScanForObstacle(entity, state.clockwise == MoveStateClockwise::Clockwise, transform, wallHug);
+}
+
+/// Transition from one grid cell to another requires another check for obstacle in the line
+template <MoveState S>
+void HandleCellTransition(ecs::Registry& registry)
+{
+	registry.Each<const MoveStateTagComponent<S>, WallHug, Transform>(
+	    [](entt::entity entity, const MoveStateTagComponent<S>& state, WallHug& wallHug, Transform& transform) {
+		    const auto position = glm::xz(transform.position);
+		    const auto positionId = Map::GetGridCell(position);
+		    const auto goalId = Map::GetGridCell(state.stepGoal);
+		    if (positionId != goalId)
+		    {
+			    CellTransition(entity, state, transform, wallHug);
+		    }
+	    });
+}
+
+// TODO: move_map_object is more complex than this. update to the map might be needed
```

Comment:
What is move_map_object? RE'd function?

Reply:
This is a carry-over from my python implementation that was based on RE